### ISSUE TITLE
Salt formulas should be considered as a new root

### ISSUE THREAD

**Issue by kiorky** [saltstack/salt#6739]
Today, it is impossible to use a new manually checkouted formula without having to declare it in the master file_roots. The behavior indicated on https://salt.readthedocs.org/en/latest/topics/conventions/formulas.html#adding-a-formula-directory-manually is wrong as it is then more that dropping a directory inside your salt states's root.

To fix that, when we use formulas as a local checkout, the top level directory should be stripped off of the resolution.

eg: if you have /srv/salt/salt-formula/salt/init.sls
Other states can then "include salt" rather having to include "salt-formula.salt"

Evidently, without, the fileserver can not resolve the "salt" directory inside the "formula checkouted directory" when you would have used "include .saltcheckoutdir.salt"
Indeed, it works for gitfs_remote, but it's because the remote is considered as a new root strippening then the offending top directory.

To sum up, I think the behavior with a local checkout of a formula and using it as a gitfs_remote should be the same and local checkouts should be picken up without any further configuration upon the initial checkout.

Related to https://github.com/saltstack-formulas/salt-formula/issues/4
@thatch45 @basepi @AntoineCezar

**Comment by aboe76** [comment]
@akoumjian I don't see the symlink at the moment am I missing something?

So far I can do the following:
spm build /home/aboe/Development/git/zsh-formula/

which creates /srv/spm/zsh-201506-1.spm

spm info zsh:
```
Name: zsh
Version: 201506
Release: 1
Install Date: Tue, 30 Jun 2015 21:23:38 GMT
Supported OSes: RedHat, Debian, Suse, Arch
Supported OS families: RedHat, Debian, Suse, Arch
Dependencies: None
OS Dependencies: None
OS Family Dependencies: None
Summary: Formula to install zsh
Description:
Formula to install zsh
```
spm files zsh
```
/srv/salt/zsh
/srv/salt/zsh/LICENSE
/srv/salt/zsh/zsh
/srv/salt/zsh/zsh/map.jinja
/srv/salt/zsh/zsh/init.sls
/srv/salt/zsh/zsh/defaults.yaml
/srv/salt/zsh/zsh/completions.sls
/srv/salt/zsh/zsh/grml
/srv/salt/zsh/zsh/grml/init.sls
/srv/salt/zsh/zsh/grml/repo.sls
/srv/pillar/zsh.sls.orig
/srv/salt/zsh/README.rst
/srv/salt/zsh/FORMULA
```

and then installing it:
spm local install /srv/spm/zsh-201506-1.spm
Installing package from file /srv/spm/zsh-201506-1.spm
Proceed? [N/y] y
... installing

/srv/salt/ # tree zsh 
```
zsh
├── FORMULA
├── LICENSE
├── README.rst
└── zsh
    ├── completions.sls
    ├── defaults.yaml
    ├── grml
    │   ├── init.sls
    │   └── repo.sls
    ├── init.sls
    └── map.jinja
```

shouldn't it be:
```
/srv/salt # tree zsh 
zsh
├── completions.sls
├── defaults.yaml
├── grml
│   ├── init.sls
│   └── repo.sls
├── init.sls
└── map.jinja
```

But than again you will be missing the README.rst and LICENSE file...

**Comment by akoumjian** [comment]
I believe this is not my version of spm, tech hat did a rewrite and that is
the one in the pr

**Comment by techhat** [comment]
Yes, this is a complete rewrite.

**Comment by aboe76** [comment]
@techhat any Idea on how to tackle this, I would like to see spm install a package that is readily available to use in top.sls referencing the package name zsh instead of zsh.zsh.

samething with the naming convention spm package zsh, while it is realy zsh-formula on gitfs?

**Comment by xlotlu** [comment]
How about supporting wildcards? This feels straightforward and natural:

```
file_roots:
  base:
    - /srv/salt
    - /srv/salt_formulas/*
```

**Comment by lorengordon** [comment]
@xlotlu, the wildcard approach doesn't solve the problem that kiorky mentioned in the OP of how to reference the state...

Give this file_roots definition:
```
file_roots:
  base:
    - /srv/salt
    - /srv/salt_formulas/*
```
And this folder structure:
```
/srv/salt
/srv/salt_formulas/mysql-formula/mysql/init.sls
/srv/salt_formulas/php-formula/php/init.sls
```
The resulting state references would be `mysql-formula.mysql` and `php-formula.php`. The desired state references are just `mysql` and `php`.

**Comment by xlotlu** [comment]
@lorengordon excuse me if I'm missing something obvious. Given your directory structure, the state references you describe would result only if there wasn't a wildcard:
```
file_roots:
  base:
    - /srv/salt
    - /srv/salt_formulas
```
With a wildcard, the internal representation after parsing the glob would be just as if one defined:
```
file_roots:
  base:
    - /srv/salt
    - /srv/salt_formulas/mysql-formula
    - /srv/salt_formulas/php-formula
```
making the `mysql` and `php` states accessible directly as desired.

**Comment by lorengordon** [comment]
@xlotlu, ahh, no, I'm the one missing the obvious. That implementation is perfectly clear when explained with the example like that.

**Comment by ghostsquad** [comment]
Is this problem solved with the officially supported/created Salt Package Manager (SPM)?
https://docs.saltstack.com/en/latest/topics/spm/master.html#installing-packages
